analyzers/src/SonarAnalyzer.CSharp/Rules/NonDerivedPrivateClassesShouldBeSealed.cs

```diff
@@ -24,48 +24,80 @@ namespace SonarAnalyzer.Rules.CSharp
     public sealed class NonDerivedPrivateClassesShouldBeSealed : SonarDiagnosticAnalyzer
     {
         private const string DiagnosticId = "S3260";
-        private const string MessageFormat = "Private classes or records which are not derived in the current assembly should be marked as 'sealed'.";
+        private const string MessageFormat = "{0} {1} which are not derived in the current assembly should be marked as 'sealed'.";
 
         private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
-
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
+        private static readonly ImmutableHashSet<SyntaxKind> KindsToBeDescended = ImmutableHashSet.Create(
+            SyntaxKind.CompilationUnit,
+            SyntaxKind.NamespaceDeclaration,
+            SyntaxKindEx.FileScopedNamespaceDeclaration,
+            SyntaxKind.InterfaceDeclaration,
+            SyntaxKind.ClassDeclaration,
+            SyntaxKind.StructDeclaration,
+            SyntaxKindEx.RecordClassDeclaration,
+            SyntaxKindEx.RecordStructDeclaration);
+
+        private static readonly ImmutableHashSet<SyntaxKind> PossiblyVirtualKinds = ImmutableHashSet.Create(
+            SyntaxKind.MethodDeclaration,
+            SyntaxKind.PropertyDeclaration,
+            SyntaxKind.EventDeclaration,
+            SyntaxKind.IndexerDeclaration);
+
         protected override void Initialize(SonarAnalysisContext context) =>
-            context.RegisterNodeAction(c =>
+            context.RegisterTreeAction(c =>
             {
-                var typeDeclarationSyntax = (TypeDeclarationSyntax)c.Node;
-                if (!c.IsRedundantPositionalRecordContext()
-                    && IsPrivateButNotSealedType(typeDeclarationSyntax)
-                    && !HasVirtualMembers(typeDeclarationSyntax))
+                var declarations = c.Tree.GetRoot().DescendantNodes(x => x.IsAnyKind(KindsToBeDescended))
+                     .Where(x => x.IsAnyKind(SyntaxKind.ClassDeclaration, SyntaxKindEx.RecordClassDeclaration))
+                     .Select(x => (TypeDeclarationSyntax)x);
```

Comment:
Main idea: working with symbols is expensive and we want to delay that and call it only if really needed. Any syntax level filtering we do before working with the semantic model is a gain in performance.

I would move `.Where(x => !IsSealed(x));` here, before the SemanticModel calls.
The same with `!HasVirtualMembers(x)`.
As a further improvement, you could do a preliminary check for the modifiers too (see `PrivateKeyword`, `FileKeyword` from `IsPossiblyDerived`).

Then, you can also check if the collection has any declarations to avoid retrieving the model and symbols if there is no need.

Reply:
Two points about this:

- I can move `IsSealed` check but not the  `HasVirtualMembers` check in the `declarations` LINQ query. The reason for this is that if a class does have the virtual keyword, even though It can't be `sealed`, it might derive from some other class on the tree, so I need it in my list to do the `.DerivesFrom` check afterwards.   

Example:

```csharp
        file class A { } // can't be sealed because of "B"
        file class B:A { public virtual void Foo() { } } // obviously can't be sealed
```
I am never raising on `B` since it has a virtual member, but in order to understand that I should _not_ raise on A, I need to know that `B` derives from it.

- If there is not at least one "candidate", meaning a: `file` or `private` classes/records that does not have virtual members and is not static/sealed/abstract, the SemanticModel is never invoked. That's the reason I wrapped it in a Lazy, it is evaluated at the last possible moment:

```csharp
if (!IsSealed(declaration)   // No SemanticModel
&& !HasVirtualMembers(declaration) // No SemanticModel
&& !IsPossiblyDerived(declaration, model, symbols, out var modifierDescription)) // SemanticModel here
```
And inside `IsPossibleDerived`:

```csharp
private static bool IsPossiblyDerived(...)
{
if (declaration.Modifiers.Any(SyntaxKind.PrivateKeyword))
{
    // Evaluate SemanticModel
    // DoStuff
    return true/false;
}
if (declaration.Modifiers.Any(SyntaxKindEx.FileKeyword))
{
    // Evaluate SemanticModel
    // DoStuff
    return true/false;
}

return true;
}

```

Reply:
Let's have a call and try to take a second look at it together. I think it's worth taking some time to check if there are any declarations left after filtering and also move the syntax-level conditions before accessing the semantic model.